lib/ansible/modules/clustering/k8s/k8s_auth.py

```diff
@@ -0,0 +1,278 @@
+    os_oauth = OAuth2Session(client_id='openshift-challenging-client')
+    authorization_url, state = os_oauth.authorization_url(auth_endpoint, state="1", code_challenge_method='S256')
+
+    basic_auth_header = make_headers(basic_auth='{}:{}'.format(username, password)).get('authorization')
+
+    # Request authorization code using basic auth credentials
+    challenge_response = os_oauth.get(
+        authorization_url,
+        headers={'X-Csrf-Token': state, 'authorization': basic_auth_header},
+        verify=verify_ssl,
+        allow_redirects=False
+    )
+
+    # TODO: check if 401 and fail
+
+    # In here we have `code` and `state`, I think `code` is the important one
+    qwargs = {k: v[0] for k, v in parse_qs(urlparse(challenge_response.headers['Location']).query).items()}
+    qwargs['grant_type'] = 'authorization_code'
+
+    # Using authorization code given to us in the Location header of the previous request, request a token
+    auth = os_oauth.post(
+        token_endpoint,
+        headers={
+            'Accept': 'application/json',
+            'Content-Type': 'application/x-www-form-urlencoded',
+            # This is just base64 encoded 'openshift-challenging-client:'
+            'Authorization': 'Basic b3BlbnNoaWZ0LWNoYWxsZW5naW5nLWNsaWVudDo='
+        },
+        data=urlencode(qwargs),
+        verify=verify_ssl
+    ).json()
+
+    # TODO: do some checks here as well and fail? 
+
+    return auth
+
+
+def openshift_logout(host, api_key, verify_ssl):
+    url = '{}/apis/oauth.openshift.io/v1/oauthaccesstokens/{}'.format(host, api_key)
+    headers = {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer {}'.format(api_key)
+    }
+    json = {
+            "apiVersion": "oauth.openshift.io/v1",
+            "kind": "DeleteOptions"
+    }
+
+    requests.delete(url, headers=headers, json=json, verify=verify_ssl)
+
+    # TODO: check return value and fail
+
+
+def execute_module(module):
+    state = module.params.get('state')
+    host = module.params.get('host')
+    username = module.params.get('username')
+    password = module.params.get('password')
+    verify_ssl = module.params.get('verify_ssl')
+    ssl_ca_cert = module.params.get('ssl_ca_cert')
+    api_key = module.params.get('api_key')
+
+    # python-requests takes either a bool or a path to a ca file as the 'verify' param
+    if verify_ssl and ssl_ca_cert:
+        ca_verify = ssl_ca_cert
+    else:
+        ca_verify = verify_ssl
+
+    # Get needed info to access authorization APIs
+    oauth_info = requests.get('{}/.well-known/oauth-authorization-server'.format(host),
+                                                                verify=verify_ssl).json()
+#    auth_endpoint = oauth_info['authorization_endpoint']
+#    token_endpoint = oauth_info['token_endpoint']
+    # TODO: remove DEBUG entries
+    auth_endpoint = f'{host}/oauth/authorize'
```

Comment:
Are f-strings available in Python2?

Reply:
This was only a thing I'm using for debugging to get the whole auth procedure through mitmproxy. I'll remove it anyways.